ci-operator/step-registry/baremetal/lab/pre/load-balancer/baremetal-lab-pre-load-balancer-commands.sh

```diff
@@ -1,2 +1,202 @@
+    bind *:443
+    mode tcp
+    balance source
+$(
+  for bmhost in $(yq e -o=j -I=0 '.[]' hosts.yaml); do
+    # shellcheck disable=SC1090
+    . <(echo "$bmhost" | yq e 'to_entries | .[] | (.key + "=\"" + .value + "\"")')
+    # shellcheck disable=SC2154
+    if [[ $name == *worker* ]]; then
+    echo "    server $name $ip:443 check inter 1s"
+    fi
+  done
+)
+EOF
+
+# Generate dhclient.conf
+cat > "${CONF_DIR}/dhclient.conf" << EOF
+# Configuration file for /sbin/dhclient.
+#
+# This is a sample configuration file for dhclient. See dhclient.conf's
+#       man page for more information about the syntax of this file
+#       and a more comprehensive list of the parameters understood by
+#       dhclient.
+#
+# Normally, if the DHCP server provides reasonable information and does
+#       not leave anything out (like the domain name, for example), then
+#       few changes must be made to this file, if any.
+#
+
+option rfc3442-classless-static-routes code 121 = array of unsigned integer 8;
+
+send host-name = gethostname();
+request subnet-mask, broadcast-address, time-offset, host-name,
+        netbios-name-servers, netbios-scope, interface-mtu,
+        rfc3442-classless-static-routes, ntp-servers;
+
+# Assuming eth1 will be the interface with the default gateway route
+interface "eth1" {
+    also request routers, domain-name, domain-name-servers, domain-search,
+        dhcp6.name-servers, dhcp6.domain-search, dhcp6.fqdn, dhcp6.sntp-servers;
+}
+
+#send dhcp-client-identifier 1:0:a0:24:ab:fb:9c;
+#send dhcp-lease-time 3600;
+#supersede domain-name "fugue.com home.vix.com";
+#prepend domain-name-servers 127.0.0.1;
+#require subnet-mask, domain-name-servers;
+#timeout 60;
+#retry 60;
+#reboot 10;
+#select-timeout 5;
+#initial-interval 2;
+#script "/sbin/dhclient-script";
+#media "-link0 -link1 -link2", "link0 link1";
+#reject 192.33.137.209;
+
+#alias {
+#  interface "eth0";
+#  fixed-address 192.5.5.213;
+#  option subnet-mask 255.255.255.255;
+#}
+
+#lease {
+#  interface "eth0";
+#  fixed-address 192.33.137.200;
+#  medium "link0 link1";
+#  option host-name "andare.swiftmedia.com";
+#  option subnet-mask 255.255.255.0;
+#  option broadcast-address 192.33.137.255;
+#  option routers 192.33.137.250;
+#  option domain-name-servers 127.0.0.1;
+#  renew 2 2000/1/12 00:00:01;
+#  rebind 2 2000/1/12 00:00:01;
+#  expire 2 2000/1/12 00:00:01;
+#}
+
+EOF
+
+timeout -s 9 180m ssh "${SSHOPTS[@]}" "root@${AUX_HOST}" bash -s -- \
+  "${NAMESPACE}" "${IPI}" << 'EOF'
+set -o nounset
+set -o errexit
+set -o pipefail
+set -o allexport
+
+BUILD_USER=ci-op
+BUILD_ID="${1}"
+IPI="${2}"
+set +o allexport
+
+# shellcheck disable=SC2174
+mkdir -m 755 -p "/var/builds/$BUILD_ID/haproxy"
+EOF
+
+echo "Uploading the haproxy.cfg file to the auxiliary host..."
+scp "${SSHOPTS[@]}" "${CONF_DIR}/haproxy.cfg" "root@${AUX_HOST}:/var/builds/$NAMESPACE/haproxy"
+scp "${SSHOPTS[@]}" "${CONF_DIR}/dhclient.conf" "root@${AUX_HOST}:/var/builds/$NAMESPACE/haproxy"
```

Comment:
I suppose you are still working on the other steps that needs porting here (docker commands and similar).

I would suggest reducing both the number of ssh connections we make and minimizing the amount of remote instructions. Try to make a tradeoff and balance as most as you can.

Two scp will need twice the time than 1 copying both the (small) files and, with the ssh connection earlier to create a directory, we need three times the minimum time needed to reach the server once  and (1) create the directory, (2) push the two files, e.g., from env variables passed via arguments.

Reply:
Sure, the rollback will be in this PR

As for minimizing scp/ssh connections, WDYT about:

Moving the logic under one ssh connections by using multiple heredocs (EOFs).
EOF will be the main one. Nested heredocs to be added as needed like EOF!, EOF2 and so on 

This way, only 1 ssh connections Is needed and 1 scp for copying hosts.yaml to target machine

I have testes nested heredocs and it seems to be working.

Reply:
Multiple heredocs are a bit difficult in terms of readability. I'd also keep on having as most logic as possible executing on the pods, rather than on the remote host.

What I'm doing is (1) filling a variable rather than a file and (2) passing the content of the variable as an argument to the remote script.

> 1 scp for copying hosts.yaml to target machine

The hosts.yaml file is expected to be in the /var/builds/$BUILD_ID folder right after the execution of the reserve steps.

Reply:
For the sake of readability, heredocs can be renamed as needed

>  I'd also keep on having as most logic as possible executing on the pods, rather than on the remote host.

Sorry, i did not got this one. commands are either execute on Jenkins agent or remote host. do you mean Jenkins pod ?


As for passing a var as an arg, substitution inside vat for below code yields errors:

$(
for bmhost in $(yq e -o=j -I=0 '.[] | select(.name == "master*")' hosts.yaml); do
  # shellcheck disable=SC1090
  . <(echo "$bmhost" | yq e 'to_entries | .[] | (.key + "=\"" + .value + "\"")')
  # shellcheck disable=SC2154
  echo "    server $name $ip:6443 check inter 1s"
done
)

Reply:
They are ocp pods running in the prow build farm ocp clusters, created via the ci-operator.

Get a look at the dhcp pxe step for reference about a for loop.